### PR TITLE
Make FORM auth work with JPA security

### PR DESCRIPTION
**Pull request by stuartwdouglas**

### CODE REVIEW

**Comment by gsmet** [comment]
@stuartwdouglas is this something you wanted for 1.4?

File: extensions/security-jpa/deployment/src/test/java/io/quarkus/security/jpa/FormAuthJpaTestCase.java [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+
+    private static final String APP_PROPS = "" +
+            "quarkus.datasource.db-kind=h2\n" +
+            "quarkus.datasource.username=sa\n" +
+            "quarkus.datasource.password=sa\n" +
+            "quarkus.datasource.jdbc.url=jdbc:h2:mem:minimal-config'\n" +
+            "quarkus.hibernate-orm.sql-load-script=import.sql\n" +
+            "quarkus.hibernate-orm.database.generation=drop-and-create\n" +
+            "#quarkus.hibernate-orm.log.sql=true\n" +
+            "quarkus.http.auth.form.enabled=true\n" +
```
Comment (FroMage) [inline extensions/security-jpa/deployment/src/test/java/io/quarkus/security/jpa/FormAuthJpaTestCase.java:29]: Are these all really required? https://quarkus.io/guides/security#form-based-authentication says they all have a default, except the encryption key, but IMO even that one could default to one generated at startup.

**Review by FroMage — APPROVED** [review]
LGTM, but I don't quite understand why we need both an `IdentityProvider` and a `TrustedIdentityProvider`? It appears that the first is only used by Basic Auth, while the second only by Form Auth?

FTR, I don't think this is what @danielpetisme was asking for the other day. IIRC what he was asking for is the ability to have an endpoint accepting auth (basic/form or just providing user/pass as custom payload) and returning a JWT that grants access to the rest of the API. 

For example, this was a supported use-case in redpipe: https://github.com/FroMage/redpipe/blob/master/redpipe-example-wiki-keycloak-jooq/src/main/java/net/redpipe/example/wiki/keycloakJooq/ApiResource.java#L57

```java
@NoAuthRedirect // this resource is not web-based so no redirect on auth missing/fail
@RequiresUser // require a user for almost every endpoint
@Produces(MediaType.APPLICATION_JSON)
@Path("/wiki/api")
public class ApiResource {
	
	@NoAuthFilter // except here, we want no auth filter: don't require a user
	@Produces("text/plain")
	@GET
	@Path("token")
	public Single<Response> token(@HeaderParam("login") String username, 
			@HeaderParam("password") String password,
			@Context JWTAuth jwt,
			@Context AuthProvider auth){
		
		JsonObject creds = new JsonObject()
				.put("username", username)
				.put("password", password);
		return fiber(() -> {
			User user;
			try {
				user = await(auth.rxAuthenticate(creds));
			}catch(VertxException x) {
				return Response.status(Status.FORBIDDEN).build();
			}
			
			boolean canCreate = await(user.rxIsAuthorised("create"));
			boolean canUpdate = await(user.rxIsAuthorised("update"));
			boolean canDelete = await(user.rxIsAuthorised("delete"));
			JsonArray permissions = new JsonArray();
			if(canCreate)
				permissions.add("create");
			if(canUpdate)
				permissions.add("update");
			if(canDelete)
				permissions.add("delete");
			
	        String jwtToken = jwt.generateToken(
	        		new JsonObject()
	        		.put("username", username)
	        		.put("permissions", permissions),
	                new JWTOptions()
	                  .setSubject("Wiki API")
	                  .setIssuer("Vert.x"));
	        return Response.ok(jwtToken).build();
		});
	}
```

As you can see, this endpoint has no security filter, and gets the user/pass however it wants from the user, then delegates auth to however it was set up (jpa possibly), and creates a JWT token as response, that the API user will then use for auth in the subsequent requests. I'm not sure it's possible to do that ATM in Quarkus, but perhaps I'm wrong?

Now, perhaps this PR is unrelated, and I'm off-topic. If so, sorry.

File: extensions/security-jpa/deployment/src/test/java/io/quarkus/security/jpa/FormAuthJpaTestCase.java (hunk 1 above)
Comment (stuartwdouglas) [reply extensions/security-jpa/deployment/src/test/java/io/quarkus/security/jpa/FormAuthJpaTestCase.java:29]: I just copied from the existing test, you don't actually need all of them.

**Comment by stuartwdouglas** [comment]
The reason we need this is that at the moment we only encode the username in the cookie, rather than the username + roles. The advantage of this approach is that it falls back to the extension that originally created the SecurityIdentity to recreated it, but I don't know how important that is, and if it is important then they could do the same thing with a SecurityIdentityAugmentor. This would mean that subsequent requests could avoid the database entirely as the username and roles are in an encrypted cookie.

I think I will merge this but look at improving this situation in a subsequent PR. 

With regards to the rest endpoint this is tangentially related, we need this in order to support persistent logins with our encrypted cookies.

File: extensions/security-jpa/deployment/src/test/java/io/quarkus/security/jpa/FormAuthJpaTestCase.java (hunk 1 above)
Comment (FroMage) [reply extensions/security-jpa/deployment/src/test/java/io/quarkus/security/jpa/FormAuthJpaTestCase.java:29]: Hah, OK.

**Comment by FroMage** [comment]
> The reason we need this is that at the moment we only encode the username in the cookie, rather than the username + roles. 

OK, I had missed that, I didn't see it didn't check for password. So for login we call `IdentityProvider` and then we get a session cookie and at every usage of the cookie we use the `TrustedIdentityProvider`?

I kinda feel like those two methods should be part of the same interface, no? This way every provider who knows how to fetch login data can authenticate and load user data.

I agree the JWT issue is orthogonal.

**Comment by hosseinmp76** [comment]
I have a problem using jpa and form auth. whene I use correct username and password I get 200 when I use incorrect password I get 401 but when I use incorrect user name I get no response.
Is this the designed behaviour?
I uses version 1.5.1
surprisingly when I use debug log level I get this at last line:
DEBUG [io.qua.ver.htt.run.sec.PersistentLoginManager] (executor-thread-1) The new cookie will expire at  ...

**Comment by webfolderio** [comment]
@stuartwdouglas @hosseinmp76 This problem still exist. If username is not found, request hangs indefinitely.

[FormAuthJpaTestCase.java](https://github.com/quarkusio/quarkus/blob/master/extensions/security-jpa/deployment/src/test/java/io/quarkus/security/jpa/FormAuthJpaTestCase.java) tests only happy path. It will nice to test if username is not exist.

If username is not found, generated byte code returns null. It will better to throw exception instead of null.
[QuarkusSecurityJpaProcessor.java#L172](https://github.com/quarkusio/quarkus/blob/master/extensions/security-jpa/deployment/src/main/java/io/quarkus/security/jpa/deployment/QuarkusSecurityJpaProcessor.java#L172)

**Pull request merged**